File: docs/practice/practice.py
"""Practice file for nvim-dev drills.

Intentionally messy and repetitive so you have things to navigate, refactor,
rename, and reformat. See docs/10-practice-drills.md.
"""

import math


# TODO: rename this function to `area_of_circle` using <leader>rn (Drill 16)
def calc(r):
    return math.pi * r * r


def calc_volume(r, h):
    # FIXME: this uses calc() — jump to its definition with gd (Drill 15)
    return calc(r) * h


class shape:
    def __init__(self, name, sides):
        self.name = name
        self.sides = sides

    def describe(self):
        print("describe something w")
        return self.name + " has " + str(self.sides) + " sides"


# Three similar lines — great for multi-cursor / macros (Drill 19)
sq = shape("square", 4)
tri = shape("triangle", 3)
pent = shape("pentagon", 5)

# A list to turn into function calls with a macro (Drill 20)
# alpha
# beta
# gamma

values = [1, 2, 3, 4, 5]  # badly formatted: fix with <leader>cf (Drill 18)
total = 0
for v in values:
    total = total + v

print("total:", total)
print(calc(2))
print(calc_volume(2, 10))
print(sq.describe())
print(tri.describe())
print(pent.describe())

File: docs/practice/practice.cpp
// Practice file for nvim-dev drills (C++).
// Messy on purpose: navigate, rename, reformat, and refactor.
// See docs/10-practice-drills.md.

#include <iostream>
#include <string>
#include <vector>

// TODO: rename `add` to `sum_ints` with <leader>rn (Drill 16)
int add(int a, int b) { return a + b; }

int multiply(int a, int b) {
  return a * b;
} // badly formatted: <leader>cf (Drill 18)

struct Point {
  double x;
  double y;
};

// Put the cursor on `Point` above and press gd from its use below (Drill 15)
double distance(const Point &p1, const Point &p2) {
  double dx = p1.x - p2.x;
  double dy = p1.y - p2.y;
  return std::sqrt(dx * dx + dy * dy); // hover std::sqrt with K
}

int main() {
  std::vector<int> nums = {1, 2, 3, 4, 5};
  int total = 0;
  for (int n : nums) {
    total = add(total, n);
  }

  // Three similar lines: try multi-cursor (<C-n>) to rename `pt` -> `pos`
  Point pt1 = {0.0, 0.0};
  Point pt2 = {3.0, 4.0};
  Point pt3 = {6.0, 8.0};
  Point pt4 = {6.1, 2.3};

  std::cout << "total = " << total << std::endl;
  std::cout << "product = " << multiply(6, 7) << std::endl;
  std::cout << "dist = " << distance(pt1, pt2) << std::endl;
  std::cout << "dist = " << distance(pt2, pt3) << std::endl;
  return 0;
}
